internal/staging/primitives/subsidy.go

```diff
@@ -0,0 +1,458 @@
+// required when calculating block and vote subsidies.  These values are
+// typically well-defined and unique per network.
+type SubsidyParams interface {
+	// BlockOneSubsidy returns the total subsidy of block height 1 for the
+	// network.  This is separate since it encompasses the initial coin
+	// distribution.
+	BlockOneSubsidy() int64
+
+	// BaseSubsidyValue returns the starting base max potential subsidy amount
+	// for mined blocks.  This value is reduced over time and then split
+	// proportionally between PoW, PoS, and the Treasury.  The reduction is
+	// controlled by the SubsidyReductionInterval, SubsidyReductionMultiplier,
+	// and SubsidyReductionDivisor parameters.
+	//
+	// NOTE: BaseSubsidy must be a max of 140,739,635,871,744 atoms or incorrect
+	// results will occur due to int64 overflow.  This value comes from
+	// MaxInt64/MaxUint16 = (2^63 - 1)/(2^16 - 1) = 2^47 + 2^31 + 2^15.
+	BaseSubsidyValue() int64
+
+	// SubsidyReductionMultiplier returns the multiplier to use when performing
+	// the exponential subsidy reduction described by the CalcBlockSubsidy
+	// documentation.
+	SubsidyReductionMultiplier() int64
+
+	// SubsidyReductionDivisor returns the divisor to use when performing the
+	// exponential subsidy reduction described by the CalcBlockSubsidy
+	// documentation.
+	SubsidyReductionDivisor() int64
+
+	// SubsidyReductionIntervalBlocks returns the reduction interval in number
+	// of blocks.
+	SubsidyReductionIntervalBlocks() int64
+
+	// StakeValidationBeginHeight returns the height at which votes become
+	// required to extend a block.  This height is the first that will be voted
+	// on, but will not include any votes itself.
+	StakeValidationBeginHeight() int64
+
+	// VotesPerBlock returns the maximum number of votes a block must contain to
+	// receive full subsidy once voting begins at StakeValidationBeginHeight.
+	VotesPerBlock() uint16
+}
+
+// totalProportions is the sum of the proportions for the subsidy split between
+// PoW, PoS, and the Treasury.
+//
+// This value comes from the fact that the subsidy split is as follows:
+//
+// Prior to DCP0010: PoW: 6, PoS: 3, Treasury: 1 => 6+3+1 = 10
+//    After DCP0010: PoW: 1, PoS: 8, Treasury: 1 => 1+8+1 = 10
+//
+// Therefore, the subsidy split percentages are:
+//
+// Prior to DCP0010: PoW: 6/10 = 60%, PoS: 3/10 = 30%, Treasury: 1/10 = 10%
+//    After DCP0010: PoW: 1/10 = 10%, PoS: 8/10 = 80%, Treasury: 1/10 = 10%
+const totalProportions = 10
+
+// SubsidyCache provides efficient access to consensus-critical subsidy
+// calculations for blocks and votes, including the max potential subsidy for
+// given block heights, the proportional proof-of-work subsidy, the proportional
+// proof of stake per-vote subsidy, and the proportional treasury subsidy.
+//
+// It makes using of caching to avoid repeated calculations.
+type SubsidyCache struct {
+	// The following fields are protected by the mtx mutex.
```

Comment:
Since this comment is above `mtx` which implicitly references it would it not be better to say `This mutex protects the  following fields:`?

Reply:
Sounds good to me.  Changed.